### ISSUE TITLE
MJPG stream connection count imbalance

### ISSUE THREAD

**Issue by sgstair** [Motion-Project/motion#1535]
I have a seemingly reliable repro for apparently a connection reference count problem in latest motion x86 (4.4.0)
My system is 5 netcams (rtsp) with motioneye managing the motion configuration. I've seen this issue 4 or 5 times now (once intentionally)
If I:
1) Disconnect a netcam from the network
2) Restart motioneye (restarts motion)
    *I'm leaving a web browser pointed to the motioneye page, and motioneye is as a result proxying the MJPG preview stream from motion*
    *At this point I see the "unable to open video device" display on the MJPG preview stream for the disconnected camera*
3) Idle for some time (1-2 hours, not sure exactly if there is a time requirement)
4) Reconnect the netcam
    *Very shortly after, motion recognizes the netcam, continues to take scheduled pictures and capture video on movement*
    **However, the MJPG stream continues to stay as the grey "unable to open video device" screen, and does not update**

Further:
* Manually connect a browser to the MJPG stream motioneye is connecting to
    *The manually connected stream does correctly update and show the current live feed*
    **While the browser is connected, the feed also starts to show the live feed in the motioneye proxied stream, this suggests to me a `cnct_count` reference loss**
* Upon closing the browser, the motioneye MJPG stream view again stops updating (`cnct_count` has probably returned to 0, despite motioneye still being connected)

I saw one mechanism that could cause this, but I don't see any evidence for it happening in the logs.
`webu_mhd_deinit` may decrease the connection count where the count was not incremented; it looks like this can only happen in the mjpg path if `webu_stream_checks` fails, though this is a denial of service scenario potentially?
The count should perhaps be incremented in `webu_answer_strm_type` (because after this point, the connection being closed will result in the count being decremented) rather than being incremented in `webu_stream_mjpeg` calling `webu_stream_cnct_count`.

I've been tracking this in https://github.com/motioneye-project/motioneye/issues/2429

It's been busy lately but I'll try to get more verbose logs today.

0. Reviewed guide and contributing documents? (Yes/No): yes
1. version [x.y.z, hash, other]: 4.4.0
2. installed as a package or compiled from sources [deb, rpm, git, other]: deb
3. standalone or part of third party [motion, MotionEyeOS, other]:  motioneye
4. video stream source [V4L (card or USB), net cam (mjpeg, rtsp, other), mmal]: rtsp
5. hardware [x86, ARM, other]: x86 (Intel(R) Celeron(R) J4125)
6. operating system [16.04, Stretch, etc, FreeBSD, other]: Debian 11.3 (bullseye)

Summarized logs related to above repro:
```
 sort -u /var/log/motion.log
[0:motion] [ALR] [ALL] conf_cmdparse: "netcam_keepalive" replaced with "netcam_params" after version 4.3.2
[0:motion] [ALR] [ALL] conf_cmdparse: "netcam_tolerant_check" replaced with "netcam_params" after version 4.3.2
[0:motion] [NTC] [ALL] config_camera: Processing camera config file camera-1.conf
[0:motion] [NTC] [ALL] config_camera: Processing camera config file camera-2.conf
[0:motion] [NTC] [ALL] config_camera: Processing camera config file camera-3.conf
[0:motion] [NTC] [ALL] config_camera: Processing camera config file camera-4.conf
[0:motion] [NTC] [ALL] config_camera: Processing camera config file camera-5.conf
[0:motion] [NTC] [ALL] conf_load: Processing thread 0 - config file /etc/motioneye/motion.conf
[0:motion] [NTC] [ALL] motion_startup: Logging to syslog
[0:motion] [NTC] [ALL] motion_startup: Motion 4.4.0 Started
[0:motion] [NTC] [ALL] motion_startup: Using default log type (ALL)
[0:motion] [NTC] [ALL] motion_startup: Using log type (ALL) log level (WRN)
[10:nc10:Back West] [ERR] [NET] Above message repeats 1 times
[10:nc10:Back West] [ERR] [NET] netcam_rtsp_handler_reconnect: norm: Reconnecting with camera....
[3:ml3:Back West] [ERR] [NET] netcam_rtsp_open_context: norm: Unable to open camera(Back West): No route to host
[3:ml3:Back West] [ERR] [VID] vid_start: Netcam RTSP failed to open
[3:ml3:Back West] [WRN] [ALL] mlp_retry: Retrying until successful connection with camera
[3:ml3:Back West] [WRN] [ALL] motion_init: Could not fetch initial image from camera
[3:ml3:Back West] [WRN] [ALL] motion_init: Motion continues using width and height from config file(s)
[5:ml5:Garage] [CRT] [NET] motion_init: Substream not available.  Image sizes not modulo 16.
[6:nc6:Front] [ERR] [NET] netcam_rtsp_handler_reconnect: norm: Reconnecting with camera....
[7:nc7:Front Door] [ERR] [NET] netcam_rtsp_handler_reconnect: norm: Reconnecting with camera....
[9:nc9:Back East] [ERR] [NET] netcam_rtsp_handler_reconnect: norm: Reconnecting with camera....
```

**Comment by Mr-Dave** [comment]
Can this be reproduced without using Motioneye?  Also, check logs at the INF or DBG level.  Historically, there was an issue with Motioneye connecting before the stream was ready and the connection counter never got incremented to tell Motion to start sending images to the port.

**Comment by sgstair** [comment]
It is pretty clear from the verbose logs, I understand better now, and it's a clear bug in motion.
I've dumped my debug-level repro log here: https://gist.github.com/sgstair/0b55ce6eb991f55af52e626015b076d4

What I didn't realize earlier is that: if the netcam is not online, the motion thread actually stops and is restarted - and the first thing it does on starting again is reinitialize the context which helpfully zeroes the `cnct_count` fields
I had incorrectly assumed this context initialization was only done once on service startup.

This means repro should be very easy and quick without motioneye:
1) Start motion trying to connect to a rtsp netcam which is disconnected / offline
2) Attach VLC/etc to the motion MJPG feed for this netcam, any time after motion start
3) Wait for motion logs to confirm motion thread restarted, and so it forgot entirely about the stream connection count
`[0:motion] [NTC] [ALL] main: Motion thread 3 restart`
4) Reattach netcam
5) Wait for motion logs to confirm motion attached to netcam
`3:ml3:Back West] [INF] [NET] netcam_rtsp_open_context: norm: Opened camera(Back West)`
6) We're now in the broken state, motion will not send data to the attached stream. If you attach a second stream, both will update normally, and then stop again if either one disconnects.

This is not a tight race, as long as motioneye / etc connects to the stream before the netcam comes online (give or take a minute due to motion's thread restarting mechanism), this bug will occur.
The thread restart occurs around every 10 seconds in my logs
```
Jun 16 09:01:43 camera motion: [0:motion] [NTC] [ALL] main: Motion thread 3 restart
Jun 16 09:01:53 camera motion: [0:motion] [NTC] [ALL] main: Motion thread 3 restart
Jun 16 09:02:03 camera motion: [0:motion] [NTC] [ALL] main: Motion thread 3 restart
Jun 16 09:02:13 camera motion: [0:motion] [NTC] [ALL] main: Motion thread 3 restart
Jun 16 09:02:23 camera motion: [0:motion] [NTC] [ALL] main: Motion thread 3 restart
```

**Comment by Mr-Dave** [comment]
Yea, so I'm not quite there at calling it a "bug".  Motion starts and does its thing when the camera comes online....now whether Motioneye picks it up, that is something different. I'm willing to look at it and see whether something is possible to accommodate MotionEye but I'm not sure a redesign of Motion would be feasible.

**Comment by sgstair** [comment]
I think this could readily happen to any higher layer using motion if the service is restarted during a camera outage.

My sense is that there's just no reason to reset `cnct_count` in `mot_stream_init` - and maybe the mutex init/deinit should be pushed back to `context_init`/`context_destroy` rather than `mot_stream_init`/`mot_stream_deinit`
The lifetime of those values and things interacting with them is apparently outside of the lifetime of the motion thread, as is very clear in this case. 

There are probably some crash race conditions too from a web request trying to interact with that mutex in the small window while the thread is stopped (due to a camera not being up when the service started)

I can only guess but it's likely more work to rework the motion loop to avoid exiting the thread when the camera isn't present on launch (though it seems to avoid exiting as long as it's successfully got a frame if the camera disappears later on)

**Comment by Mr-Dave** [comment]
I believe that this issue is resolved by the latest commit.

**Comment by sgstair** [comment]
Thanks for looking into it!
It looks like that will probably resolve the problem of the motion loop restarting before first acquiring the camera, indeed

It does seem like restarting the motion loop is still very much a normal reaction to some rare and unexpected circumstances, and doing so will still reset the stream `cnct_count`s which will unbalance them and freeze any long-term stream clients.
I suggest this is still worth fixing, but it does at least seem a lot less likely circumstance than the case this fixes

**Comment by Mr-Dave** [comment]
closed with commit a511deff214a7e40573f976bfb5178c39cb29afd
